### DISCUSSION TITLE
Is my automation (using https://github.com/upcmd/up) achieving the exact same thing as goreleaser?

### DISCUSSION THREAD

**Discussion by stephencheng** [goreleaser/goreleaser#1682] [Q&A] [answered]
Firstly, an admirer of the gorelease project here. I have no in depth knowledge of gorelease, though I wanted to use it longtime ago. 

I am the author of https://github.com/upcmd/up, which is a ci/cd automation tool, could be an alternatives/company for tools like Make, Ansible, Ant, Gradle etc.

While I was sharing my implementation of how I automated the build/tag/publish/release in reddit, I was referring to gorelease.

My post is here: https://www.reddit.com/r/devops/comments/htcswy/sharing_how_do_full_automate_build_and_publish_my/

So basically, it is just two entry tasks for me to release either all (32 distros) or (latest master daily build), and essentially to call a common shared reusable task

1. Build and publish a tagged release with 23 different possible binary distos to github project release automatically
In this task, what it does are:

* auto detect your tagged version
* auto release all distros which are supported by golang

2. A daily build and publish task to release the latest regression tested version of the osx/linux/windows platform builds only
Both of these two tasks share the same reusable common task: buildDistros

In buildDistros, it exam if a parameter/var distlist is supplied, in task 2 - daily build of latest version, it supplies the distro to be a list of: osx/linux/windows, in the publish of a tagged version, it just automatically generate a list of all support ox and archtype

Then the build also try to attempt to build all os/arch combinations. If it fails, it ignores the error and iterate through all the list. It comes down to 32 distros generated in the end.

config
https://github.com/upcmd/up/blob/master/up.yml, check the tasks: publish and Publish_latest

According to the definition of gorelease:

> The goal is to simplify the build, release and publish steps while providing variant customization options for all steps.
> 
> GoReleaser is built for CI tools, you only need to download and execute it in your build script

I think that I used UPcmd to have achieved the same result nicely. Also the customisation could be done using the orchestration workflow.

I guess UPcmd is more generic for automation, while gorelease is specialised in 'release' scenario/use case. I am just wondering if there is anything that gorelease could do better/more so that I might consider switch to invoke gorelease directly inside of the UPcmd task configuration.

**Comment by caarlos0** [comment, marked as answer]
> I guess UPcmd is more generic for automation, while gorelease is specialised in 'release' scenario/use case.

you are exactly right.

GoReleaser handles the build, packaging and releasing of projects... you can check the customization docs to see everything that's possible to do...
